[jazzy] zenoh-cpp-vendor WIP

muellerbernd opened the pull request:
I have tested the somewhat new middleware [rmw_zenoh](https://github.com/ros2/rmw_zenoh) in a local workspace. Now I try to get the corresponding nix packages running. 
When try to build them:
```bash
nix build .\#jazzy.zenoh-cpp-vendor --accept-flake-config
```
Colcon gets invoked, then colcon executes cargo and cargo cannot download stuff because it is sandboxed. According to the [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#compiling-non-rust-packages-that-include-rust-code) we would have to set the `cargoDeps`. 
I saw that in #501 was also done some work to build rust nodes.

muellerbernd commented:
Zenoh is now part of nixpkgs see [here](https://github.com/NixOS/nixpkgs/blob/62eb60588fe550b8b8f6ebe4b02d7388a5c16681/pkgs/by-name/ze/zenoh/package.nix), maybe we can use that here. I will test this in the next days.

muellerbernd commented:
now zenoh-cpp-vendor is working but zenoh-bridge-dds does not compile completely

muellerbernd commented:
update on zenoh-bridge-dds. If I check out version 1.2.1 from github it also builds, so I think a new release will fix that also.

Cakem1x commented:
Yay, thanks for your work @muellerbernd!

I was just trying to use zenoh as RMW in my devshell (via pkg `rmw-zenoh-cpp`, on distro `jazzy`) and ran into the following error using the develop branch:
```
error: builder for '/nix/store/gkazl5k3qhi1slrz6g83grk0ai4nh7p1-ros-jazzy-zenoh-cpp-vendor-0.2.1-r1.drv' failed with exit code 2;
[...]
       > [  6%] Creating directories for 'zenoh_c_vendor'
       > [ 12%] Performing download step for 'zenoh_c_vendor'
       > === ./zenoh_c_vendor (git) ===
       > Could not determine ref type of version: fatal: unable to access 'https://github.com/eclipse-zenoh/zenoh-c.git/': Could not resolve host: github.com
```

With your branch, I can successfully use rmw zenoh.

Pleune commented:
Here is my overlay to use zenoh, but I have to use a slightly out of date version from before the switch to zenoh-cpp

```nix
    rmw-zenoh-cpp = rosFinal.buildRosPackage {
      pname = "ros-${ros}-rmw-zenoh-cpp";
      inherit version;
      src = ./vendored/rmw_zenoh/rmw_zenoh_cpp;
      # Dep added directly, dont need vendor
      postPatch = ''
        sed -i '/zenoh_c_vendor/d' CMakeLists.txt
      '';
      buildType = "ament_cmake";
      nativeBuildInputs = with rosFinal; [ ament-cmake ];
      propagatedBuildInputs =
        with final;
        with rosFinal;
        with rosFinal.python3.pkgs;
        [
          zenoh-c
          rosidl-typesupport-fastrtps-c
          rosidl-typesupport-fastrtps-cpp
          rcutils
          rcpputils
          fastcdr
          ament-index-cpp
          rmw
        ];
    };

  zenoh-c = prev.stdenv.mkDerivation rec {
    name = "zenoh-c";
    pname = "zenoh-c";
    src = prev.fetchFromGitHub rec {
      name = "zenoh-c-${rev}"; # invalidate build if rev changes
      owner = "eclipse-zenoh";
      repo = "zenoh-c";
      # parse exact git commit from the zenoh_c_vendor package
      rev = builtins.elemAt (builtins.match ".*VCS_VERSION ([a-z0-9]+).*" (
        builtins.readFile ./vendored/rmw_zenoh/zenoh_c_vendor/CMakeLists.txt
      )) 0;
      hash = "sha256-4MAytgYjBqpNeeTEFRTzT3deIg4YFVE+YJnzFEJPpCE=";
    };
    cargoDeps = prev.rustPlatform.importCargoLock {
      lockFile = "${src}/Cargo.lock";
      outputHashes = {
        "zenoh-0.11.0-dev" = "sha256-UtOuH5CwOKj8ZSxnAwVlFfe5RrejAp87SN9xlkwam90=";
      };
    };
    nativeBuildInputs = with prev; [
      cmake
      rustPlatform.cargoSetupHook
      cargo
      rustc
    ];
    # Remove path building, as nix is already absolute paths
    postPatch = ''
      sed -i -e '/^preifx=.*/d' -e 's|[$]{prefix}/||' install/zenohc.pc.in
      sed -i -e 's|[$]{_IMPORT_PREFIX}/||' install/PackageConfig.cmake.in
    '';
  };
```

I hope it helps!

Pleune commented:
I have tried to update to newer versions of rmw_zenoh, but when building zenoh-cpp I get errors like below

```
error: builder for '/nix/store/dqp7haqy2vkg8dykdq90580v7wi1wnx8-ros-jazzy-rmw-zenoh-cpp-0.1.0.drv' failed with exit code 2;
       last 25 log lines:
       >   142 |   std::optional<zenoh::ext::AdvancedSubscriber<void>> sub_;
       >       |                             ^~~~~~~~~~~~~~~~~~
       > /build/rmw_zenoh_cpp/src/detail/rmw_subscription_data.hpp:142:52: error: template argument 1 is invalid
       >   142 |   std::optional<zenoh::ext::AdvancedSubscriber<void>> sub_;
       >       |                                                    ^~
       > /build/rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:57:24: error: 'ShmProvider' is not a member of 'zenoh'
       >    57 |   std::optional<zenoh::ShmProvider> & shm_provider();
       >       |                        ^~~~~~~~~~~
       > /build/rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:57:35: error: template argument 1 is invalid
       >    57 |   std::optional<zenoh::ShmProvider> & shm_provider();
       >       |                                   ^
       > /build/rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:57:24: error: 'ShmProvider' is not a member of 'zenoh'
       >    57 |   std::optional<zenoh::ShmProvider> & shm_provider();
       >       |                        ^~~~~~~~~~~
       > /build/rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp:57:35: error: template argument 1 is invalid
       >    57 |   std::optional<zenoh::ShmProvider> & shm_provider();
       >       |                                   ^
       > make[2]: *** [CMakeFiles/rmw_zenoh_cpp.dir/build.make:272: CMakeFiles/rmw_zenoh_cpp.dir/src/detail/rmw_service_data.cpp.o] Error 1
       > make[2]: *** [CMakeFiles/rmw_zenoh_cpp.dir/build.make:258: CMakeFiles/rmw_zenoh_cpp.dir/src/detail/rmw_node_data.cpp.o] Error 1
       > make[2]: *** [CMakeFiles/rmw_zenoh_cpp.dir/build.make:454: CMakeFiles/rmw_zenoh_cpp.dir/src/rmw_get_topic_names_and_types.cpp.o] Error 1
       > make[2]: *** [CMakeFiles/rmw_zenoh_cpp.dir/build.make:482: CMakeFiles/rmw_zenoh_cpp.dir/src/rmw_init.cpp.o] Error 1
       > [ 97%] Linking CXX executable rmw_zenohd
       > [ 97%] Built target rmw_zenohd
       > make[1]: *** [CMakeFiles/Makefile2:139: CMakeFiles/rmw_zenoh_cpp.dir/all] Error 2
       > make: *** [Makefile:146: all] Error 2
       For full logs, run 'nix log /nix/store/dqp7haqy2vkg8dykdq90580v7wi1wnx8-ros-jazzy-rmw-zenoh-cpp-0.1.0.drv'.
```

Pleune commented:
Related to my progress getting my overlay to work with updated rmw_zenoh https://github.com/ros2/rmw_zenoh/issues/482

muellerbernd commented:
> Related to my progress getting my overlay to work with updated rmw_zenoh [ros2/rmw_zenoh#482](https://github.com/ros2/rmw_zenoh/issues/482)

I have updated this PR to use updated version of rmw_zenoh

muellerbernd commented:
@wentasah @lopsided98 I had to reintroduce the older version of `patchAmentVendorGit` especially for `zenoh-cpp-vendor` because `zenoh-cpp-vendor` has multiple occurrences of amend_vendor and as far as I understand the new `patchAmendVendor` function does not work for that use case.

YuanYuYuan commented:
Hi @muellerbernd! I just came across your work and wanted to express my gratitude for the great contributions you've made. I can confirm that everything works perfectly on my NixOS machine. 😊

wentasah commented:
@muellerbernd Sorry for breaking your PR. I tried to extend the new `patchAmentVendor` logic to also work for `zenoh-cpp-vendor`. Can you try building this PR on top of #620?

When I tried override like this, automatic patching worked:
```nix
zenoh-cpp-vendor = lib.patchAmentVendorGit rosSuper.zenoh-cpp-vendor { };
```
It will be needed to further override the result to patch the other things that you do in this PR. If you need the git hashes, you should be able to read them from `zenoh-cpp-vendor/vendored-source.json`, however you will need to use numerical indices to choose between `zenoh-c` and `zenoh-cpp`. Let me know if you encounter any problems.

muellerbernd commented:
> @muellerbernd Sorry for breaking your PR. I tried to extend the new `patchAmentVendor` logic to also work for `zenoh-cpp-vendor`. Can you try building this PR on top of #620?
> 
> When I tried override like this, automatic patching worked:
> 
> ```nix
> zenoh-cpp-vendor = lib.patchAmentVendorGit rosSuper.zenoh-cpp-vendor { };
> ```
> 
> It will be needed to further override the result to patch the other things that you do in this PR. If you need the git hashes, you should be able to read them from `zenoh-cpp-vendor/vendored-source.json`, however you will need to use numerical indices to choose between `zenoh-c` and `zenoh-cpp`. Let me know if you encounter any problems.

I have tried to get it building in a [testing branch](https://github.com/muellerbernd/nix-ros-overlay/tree/pr-620-additions) but it is still WIP.

wentasah commented:
I'm getting `CMake Error: The source directory "/build/rmw_zenoh-release-release-jazzy-zenoh_cpp_vendor-0.2.3-1/build/zenoh_cpp_vendor-prefix/src/zenoh_cpp_vendor" does not appear to contain CMakeLists.txt.`

muellerbernd commented:
> I'm getting `CMake Error: The source directory "/build/rmw_zenoh-release-release-jazzy-zenoh_cpp_vendor-0.2.3-1/build/zenoh_cpp_vendor-prefix/src/zenoh_cpp_vendor" does not appear to contain CMakeLists.txt.`

I get the same error. Right now I have no clue what's causing this.

wentasah commented:
Hmm, it was my fault. I updated #620. With it, your overrides work.

wentasah commented:
@muellerbernd I've extracted overrides from your test branch and added them in #643 as https://github.com/lopsided98/nix-ros-overlay/commit/af4ee838e0b5d1d87fcafdf8b889a9b33464c916. I simplified the Nix expression a bit to take advantage of the fact that `vendored-source.json` is an object rather than an array (as before).

I've also removed overrides for `zenoh-bridge-dds` as they don't work with the latest overlay (in my PR). Is zenoh-bridge-dds needed nowadays? The [README says](https://github.com/eclipse-zenoh/zenoh-plugin-dds?tab=readme-ov-file#%EF%B8%8F-on-usage-with-ros-2-%EF%B8%8F) that [zenoh-plugin-ros2dds](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds) is preferable.

muellerbernd commented:
> @muellerbernd I've extracted overrides from your test branch and added them in #643 as [af4ee83](https://github.com/lopsided98/nix-ros-overlay/commit/af4ee838e0b5d1d87fcafdf8b889a9b33464c916). I simplified the Nix expression a bit to take advantage of the fact that `vendored-source.json` is an object rather than an array (as before).
> 
> I've also removed overrides for `zenoh-bridge-dds` as they don't work with the latest overlay (in my PR). Is zenoh-bridge-dds needed nowadays? The [README says](https://github.com/eclipse-zenoh/zenoh-plugin-dds?tab=readme-ov-file#%EF%B8%8F-on-usage-with-ros-2-%EF%B8%8F) that [zenoh-plugin-ros2dds](https://github.com/eclipse-zenoh/zenoh-plugin-ros2dds) is preferable.

@wentasah looks good to me. Thanks for your effort on getting also Kilted working with this overlay. @lopsided98 would be nice if you could give some feedback on this and the other mentioned PRs.
